lib/http/2/stream.rb

```diff
@@ -97,14 +98,43 @@ def receive(frame)
 
       case frame[:type]
       when :data
-        window_size = frame[:payload].bytesize
-        window_size += frame[:padding] || 0
-        @local_window -= window_size
+        # Emit DATA frame
+        frame_size = frame[:payload].bytesize
+        frame_size += frame[:padding] || 0
+        @local_window -= frame_size
         emit(:data, frame[:payload]) unless frame[:ignore]
 
-        # Automatically send WINDOW_UPDATE,
-        # assuming that emit(:data) can now receive next data
-        window_update(window_size) if window_size > 0
+        local_window_used = @local_window_max_size - @local_window
+        # If DATA frame is received with length > 0 and
+        # current received window size + delta length is strictly larger than
+        # local window size, it throws a flow control error.
+        # Note that local_window_max_size is calculated after SETTINGS ACK is
+        # received from peer, so peer must honor this limit. If the resulting
+        # local window used is strictly larger than NGHTTP2_MAX_WINDOW_SIZE,
+        # throw a flow control error too.
+        #  (https://github.com/nghttp2/nghttp2/blob/2bf3680d870953010d7e1e6e4a66510f8458cc3c/lib/nghttp2_session.c#L4905-L4922)
+        #
+        if @local_window < 0 || local_window_used > 0x7fffffff
```

Comment:
The second check is unnecessary. 
- We will already raise an error if the peer tries to negotiate a window size above max threshold
- The first part of the "or" will catch cases where peer exceeds above limit or the flow control window

Also, please drop all the text starting with "Note that..." — this is nghttp specific.